packages/plugins/persisted-queries/src/plugin.ts

```diff
@@ -0,0 +1,96 @@
+/* eslint-disable no-console */
+import { Plugin, DefaultContext } from '@envelop/core';
+import { DocumentNode, parse, Source, GraphQLError } from 'graphql';
+
+const contextProperty = 'documentId';
+
+export type PersistedQueriesStoreList = Map<string, { [key: string]: string }>;
+
+export interface PersistedQueriesStore {
+  get(): PersistedQueriesStoreList;
+}
+
+interface PluginContext {
+  [contextProperty]?: string;
+}
+
+export type UsePersistedQueriesOptions = {
+  store: PersistedQueriesStore;
+  onlyPersisted?: boolean;
+  setQueryId?: (context: Readonly<DefaultContext>) => string | undefined;
+  pickSingleList?: (context: Readonly<DefaultContext>) => string | undefined;
+};
+
+const DEFAULT_OPTIONS: Omit<UsePersistedQueriesOptions, 'store'> = {
+  onlyPersisted: false,
+};
+
+export const usePersistedQueries = (rawOptions: UsePersistedQueriesOptions): Plugin<PluginContext> => {
+  const options: UsePersistedQueriesOptions = {
+    ...DEFAULT_OPTIONS,
+    ...(rawOptions || {}),
+  };
+
+  return {
+    onParse({ context, params, extendContext, setParsedDocument }) {
+      const store = options.store.get(); // retrieve fresh instance of store Map
+      const queryId = options.setQueryId ? options.setQueryId(context) : queryIdFromSource(params.source);
```

Comment:
I think we have a similar thing in persisted-operations plugin, that @kamilkisiela added

Reply:
I think currently existing persisted-operations doesn't receive context, to begin with.
Also there is no logic to grab the id/hash from standard GraphQL source.

Probably the bit you're referring to has been introduced by @kamilkisiela in Hive rather then Envelop